File: soc-server/internal/agent_grpc/connection_manager.go
// ==============================================================================
// Package agent_grpc - Connection Manager
// File: internal/agent_grpc/connection_manager.go
// Mô tả: Quản lý tất cả kết nối gRPC active từ các Agent.
//         Thread-safe bằng sync.RWMutex vì thao tác đọc (SendCommand, lookup)
//         nhiều hơn ghi (Register, Unregister).
//
// Chức năng chính:
//   - Register/Unregister Agent connections khi stream mở/đóng
//   - SendCommand(): Tra cứu Agent → đẩy lệnh vào channel → stream gửi xuống
//   - GetOnlineAgents(): Liệt kê các Agent đang kết nối
// ==============================================================================

package agent_grpc

import (
	"fmt"
	"log"
	"sync"

	pb "soc-server/internal/generated/proto"
)

// ActiveConnection - Đại diện cho 1 kết nối gRPC active từ Agent
type ActiveConnection struct {
	AgentID    string                          // UUID của Agent
	Hostname   string                          // Hostname Endpoint
	IPAddress  string                          // IP Address
	Stream     pb.AgentService_StreamEventsServer // gRPC bidirectional stream
	CmdChan    chan *pb.CommandResponse          // Channel để đẩy lệnh xuống Agent
	Done       chan struct{}                     // Channel báo hiệu kết nối đã đóng
}

// ConnectionManager - Quản lý tập trung tất cả Agent connections
// Thread-safe: Sử dụng RWMutex cho concurrent read access
type ConnectionManager struct {
	mu          sync.RWMutex                    // Mutex bảo vệ connections map
	connections map[string]*ActiveConnection     // Map: agentID → *ActiveConnection
}

// NewConnectionManager - Khởi tạo Connection Manager mới
func NewConnectionManager() *ConnectionManager {
	return &ConnectionManager{
		connections: make(map[string]*ActiveConnection),
	}
}

// Register - Đăng ký kết nối Agent mới khi stream mở
// Nếu Agent đã có kết nối cũ, đóng kết nối cũ trước (tránh duplicate)
func (cm *ConnectionManager) Register(conn *ActiveConnection) {
	cm.mu.Lock()
	defer cm.mu.Unlock()

	// Kiểm tra xem Agent đã có kết nối cũ chưa
	if oldConn, exists := cm.connections[conn.AgentID]; exists {
		log.Printf("[CONN MANAGER] ⚠️ Agent '%s' đã có kết nối cũ, đang đóng kết nối cũ...", conn.AgentID)
		close(oldConn.Done)    // Báo hiệu goroutine cũ dừng lại
	}

	// Đăng ký kết nối mới
	cm.connections[conn.AgentID] = conn
	log.Printf("[CONN MANAGER] ✅ Đã đăng ký Agent '%s' (%s - %s). Tổng: %d agents online",
		conn.AgentID, conn.Hostname, conn.IPAddress, len(cm.connections))
}

// Unregister - Hủy đăng ký khi Agent ngắt kết nối
func (cm *ConnectionManager) Unregister(agentID string) {
	cm.mu.Lock()
	defer cm.mu.Unlock()

	if conn, exists := cm.connections[agentID]; exists {
		close(conn.Done) // Đóng channel Done
		delete(cm.connections, agentID)
		log.Printf("[CONN MANAGER] 🔴 Agent '%s' đã ngắt kết nối. Tổng: %d agents online",
			agentID, len(cm.connections))
	}
}

// SendCommand - Gửi lệnh Active Response xuống Agent cụ thể
// Tra cứu Agent theo ID → đẩy command vào channel → goroutine stream gửi xuống
//
// Trả về error nếu Agent không online hoặc channel đầy
func (cm *ConnectionManager) SendCommand(agentID string, cmd *pb.CommandResponse) error {
	cm.mu.RLock()
	conn, exists := cm.connections[agentID]
	cm.mu.RUnlock()

	if !exists {
		return fmt.Errorf("Agent '%s' không online hoặc không có kết nối gRPC active", agentID)
	}

	// Đẩy command vào channel (non-blocking)
	select {
	case conn.CmdChan <- cmd:
		log.Printf("[CONN MANAGER] 📤 Đã gửi lệnh '%s' xuống Agent '%s' (target: %s)",
			cmd.CommandType.String(), agentID, cmd.Target)
		return nil
	default:
		return fmt.Errorf("command channel của Agent '%s' đã đầy, không thể gửi lệnh", agentID)
	}
}

// GetConnection - Lấy ActiveConnection theo Agent ID
func (cm *ConnectionManager) GetConnection(agentID string) (*ActiveConnection, bool) {
	cm.mu.RLock()
	defer cm.mu.RUnlock()

	conn, exists := cm.connections[agentID]
	return conn, exists
}

// GetOnlineAgentIDs - Lấy danh sách Agent ID đang online
func (cm *ConnectionManager) GetOnlineAgentIDs() []string {
	cm.mu.RLock()
	defer cm.mu.RUnlock()

	ids := make([]string, 0, len(cm.connections))
	for id := range cm.connections {
		ids = append(ids, id)
	}
	return ids
}

// GetOnlineCount - Đếm số Agent đang online
func (cm *ConnectionManager) GetOnlineCount() int {
	cm.mu.RLock()
	defer cm.mu.RUnlock()
	return len(cm.connections)
}

// IsAgentConnected - Kiểm tra Agent có đang kết nối gRPC không
func (cm *ConnectionManager) IsAgentConnected(agentID string) bool {
	cm.mu.RLock()
	defer cm.mu.RUnlock()
	_, exists := cm.connections[agentID]
	return exists
}

File: soc-server-python/app/grpc_server/connection_manager.py
# ==============================================================================
# app/grpc_server/connection_manager.py - Quản lý kết nối gRPC active từ Agents
# Tương đương: internal/agent_grpc/connection_manager.go
# ==============================================================================

import asyncio
import logging
from typing import Dict, Optional, List

logger = logging.getLogger(__name__)


class ActiveConnection:
    """Đại diện cho 1 kết nối gRPC active từ Agent"""

    def __init__(self, agent_id: str, hostname: str, ip_address: str, context):
        self.agent_id = agent_id
        self.hostname = hostname
        self.ip_address = ip_address
        self.context = context  # gRPC ServicerContext
        # Queue để đẩy lệnh xuống Agent (tương đương CmdChan chan *pb.CommandResponse)
        self.cmd_queue: asyncio.Queue = asyncio.Queue(maxsize=100)
        # Event để báo hiệu kết nối đóng (tương đương Done chan struct{})
        self.done: asyncio.Event = asyncio.Event()


class ConnectionManager:
    """
    Quản lý tập trung tất cả Agent connections.
    Thread-safe: Sử dụng asyncio.Lock cho concurrent access.
    Tương đương: internal/agent_grpc/connection_manager.go
    """

    def __init__(self):
        self._connections: Dict[str, ActiveConnection] = {}
        self._lock = asyncio.Lock()

    async def register(self, conn: ActiveConnection):
        """Đăng ký kết nối Agent mới khi stream mở"""
        async with self._lock:
            # Nếu Agent đã có kết nối cũ → đóng cũ trước
            if conn.agent_id in self._connections:
                old = self._connections[conn.agent_id]
                logger.warning(
                    f"[CONN MANAGER] ⚠️ Agent '{conn.agent_id}' đã có kết nối cũ, đang đóng..."
                )
                old.done.set()  # Báo hiệu goroutine cũ dừng lại

            self._connections[conn.agent_id] = conn
            count = len(self._connections)
            logger.info(
                f"[CONN MANAGER] ✅ Đã đăng ký Agent '{conn.agent_id}' "
                f"({conn.hostname} - {conn.ip_address}). Tổng: {count} agents online"
            )

    async def unregister(self, agent_id: str):
        """Hủy đăng ký khi Agent ngắt kết nối"""
        async with self._lock:
            if agent_id in self._connections:
                conn = self._connections.pop(agent_id)
                conn.done.set()
                count = len(self._connections)
                logger.info(
                    f"[CONN MANAGER] 🔴 Agent '{agent_id}' đã ngắt kết nối. "
                    f"Tổng: {count} agents online"
                )

    async def send_command(self, agent_id: str, command: dict) -> bool:
        """
        Gửi lệnh Active Response xuống Agent cụ thể.
        Trả về False nếu Agent không online hoặc queue đầy.
        Tương đương: ConnectionManager.SendCommand() trong Go
        """
        async with self._lock:
            conn = self._connections.get(agent_id)

        if conn is None:
            logger.error(
                f"[CONN MANAGER] ❌ Agent '{agent_id}' không online "
                f"hoặc không có kết nối gRPC active"
            )
            return False

        try:
            conn.cmd_queue.put_nowait(command)
            logger.info(
                f"[CONN MANAGER] 📤 Đã gửi lệnh '{command.get('command_type')}' "
                f"xuống Agent '{agent_id}' (target: {command.get('target')})"
            )
            return True
        except asyncio.QueueFull:
            logger.warning(
                f"[CONN MANAGER] ⚠️ Command queue của Agent '{agent_id}' đã đầy. "
                f"Lệnh '{command.get('command_type')}' bị bỏ qua!"
            )
            return False

    async def get_connection(self, agent_id: str) -> Optional[ActiveConnection]:
        """Lấy ActiveConnection theo Agent ID"""
        async with self._lock:
            return self._connections.get(agent_id)

    async def get_online_agent_ids(self) -> List[str]:
        """Lấy danh sách Agent ID đang online"""
        async with self._lock:
            return list(self._connections.keys())

    async def get_online_count(self) -> int:
        """Đếm số Agent đang online"""
        async with self._lock:
            return len(self._connections)

    async def is_agent_connected(self, agent_id: str) -> bool:
        """Kiểm tra Agent có đang kết nối gRPC không"""
        async with self._lock:
            return agent_id in self._connections


# Singleton global instance — dùng chung giữa gRPC server và REST API
global_conn_manager = ConnectionManager()
